### DISCUSSION TITLE
Raylib+imgui project

### DISCUSSION THREAD

**Discussion by axx0** [axx0/Civ2-clone#52] [Show and tell]
Given the problems we're experiencing with EtoForms (a bunch of controls are missing and have to be added manually for win/unix/mac) I've been searching for a better library. Modern.forms would be ideal since it basically implements Winforms on mac+unix through Avalonia, but it's still at quite an early stage of development.

One drastically different library is Raylib (engine) in combination with Imgui (UI). Imgui is widely popular and is becoming something of an industry standard for interface design. Raylib is a simple library that should cover all our needs.

The difference compared to Winforms/Etoforms is substantial --> it's not an event-driven but there's a game loop. This makes at least drawing stuff much easier and allows for some animations that are hard to implement with our current engine.

I made a quick demonstration project which loads a SAV. Just change the sav name in line 15 in Main.cs.
You should get something like this.
![raylib](https://user-images.githubusercontent.com/21365802/211167982-4c0fc17b-e76a-469e-a61c-9d6cb163b342.png)

The app runs well on both my Win10 and Ubuntu systems. It takes some time to load the textures, but that can probably be optimized. Loading Civ2 WAVs seems to be a problem though.

Any feedback is appreciated.

**Comment by reubene** [comment]
Interesting I'll take a look this week

**Reply by reubene** [reply]
I've got it building and running on Win10 and my M1 mac.  I had to make a few changes and add some imports to get it working on mac but nothing structural. 

I'll PR my changes and then perhaps I might start looking into how to get the menu dialogs Going in this version

**Comment by axx0** [comment]
Imgui should have all the necessary controls for our dialogs. See what you make of it.

**Comment by reubene** [comment]
Here's where I'm at: 
![Main menu](https://user-images.githubusercontent.com/8866710/217765570-2e536bcf-d4c1-4d1f-838d-1479d0fb190a.png)
The first part of the menu is working, although I've not had a lot of luck figuring out how to style imgui to look appropriate. 

Also I'm not sure how to dictate where the imgui dialogs should show initially. 

I've also seen a few artifacts in the imgui portions on mac, I'll investigate further might be related to the styling issues. Seems to only happen on the main laptop screen not any of my secondary displays

**Reply by axx0** [reply]
Nice, I suggest you commit whatever you've achieved sofar.
I'll try to play around with Imgui some more. I was also hoping it would be easier to achieve civ2-looks. As an alternative if we hit a brick wall we could use [raygui](https://github.com/raysan5/raygui) through [this](https://github.com/NotNotTech/Raylib-CsLo).

Also I was thinking - with winforms you have a separate window for main map, status panel and minimap. All can be resized separately. In case of raylib we have to fill the entire screen with some content. We can achieve the looks of the original game and make it appear like we have 3 panels, but "resizing" these will be hard to achieve. Do we even need this feature?

**Reply by reubene** [reply]
I don't think we need the resizing feature, I don't think it was of much use, that way we can create a standard UI texture to render the other element onto and dispense with the background

**Comment by axx0** [comment]
I'll try to implement controls on my own so that we can ditch imgui. Doesn't look that hard really.

**Reply by reubene** [reply]
I was actually thinking of doing the same thing. You've already worked out a lot of the dialog sizing code in the ETO Forms side of things and we don't need a full blown widget toolkit. 

Since all dialogs are modal we can put them in a queue and pass all input to the current topmost this will disable everything else on screen which is probably useful. If the dialog is controlling it's own painting then it will know where it is on screen and it shouldn't be hard to work out where clicks were.

**Reply by reubene** [reply]
In terms of what controls we need most of them should be fairly simple, radio buttons, check boxes and buttons shouldn't be an issue. 

It gets more interesting when you get to listboxes and dropdowns that might need scrolling but I think we can cross that bridge when we get there.

I don't think the menu bar itself will be that challenging, as it's just clickable text labels on a background

**Reply by axx0** [reply]
Good thing about dialogs is that only one can be active at a time and there are no background dialogs. 
Most of the necessary logic and implementation of controls can be found [here](https://github.com/raysan5/raygui/blob/master/src/raygui.h). 
File dialogs should be the biggest problem but there's an example on raygui of how he implemented it.

**Reply by reubene** [reply]
It all depends how you want to design the file dialog. It doesn't need to be a TreeView. Or need to support autocomplete. I'm going to have a look at the marchalling mechanics for a basic dialog and then I might have a go at making a file dialog.

**Comment by crhy** [comment]
I think we should design everything to operate well on a touch screen as well.  Would be excellent to have a Civ 2 on Android in the Neo Store.

**Reply by garyrpage** [reply]
Very nice suggestion. I don't really know how portable the Raylib Engine and UI stuff in the current code would be for native mobile devices - Raylib does appear to have a makefile for Android... But from my general career experience; it might be possible to make a proof of concept that reuses the existing engine or a lightweight version of it, and then uses something like MAUI (if staying with .NET) or another mobile kit to run a native version.
